cpp/src/strings/search/contains_multiple.cu

```diff
@@ -0,0 +1,316 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/find_multiple.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cooperative_groups.h>
+#include <cub/cub.cuh>
+#include <cuda/functional>
+#include <thrust/binary_search.h>
+#include <thrust/equal.h>
+#include <thrust/fill.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/logical.h>
+#include <thrust/sequence.h>
+#include <thrust/unique.h>
+
+#include <vector>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+namespace {
+
+/**
+ * @brief Threshold to decide on using string or warp parallel functions.
+ *
+ * If the average byte length of a string in a column exceeds this value then
+ * a warp-parallel function is used.
+ */
+constexpr size_type AVG_CHAR_BYTES_THRESHOLD = 64;
+
+/**
+ * @brief Kernel for finding multiple targets in each row of input strings
+ *
+ * The d_first_bytes is sorted and unique so the d_indices and d_offsets
+ * are used to map the corresponding character to its d_targets entry.
+ *
+ * Example
+ * d_targets = ["foo", "hello", "world", "hi"]
+ *  - sorted first-chars: ['f','h','h','w']
+ * d_indices = [0, 3, 1, 2]
+ * d_first_bytes = ['f', 'h', 'w']   (unique)
+ * d_offsets = [0, 1, 3]
+ * unique_count = 3
+ *
+ * If 'h' is found, lower_bound produces pos=1 in d_first_bytes.
+ * This corresponds to d_offset[1]==1 which has two values:
+ * - (d_offsets[2] - d_offsets[1]) = (3 - 1) = 2.
+ * Set map_idx = d_offsets[1] = 1 and the two targets to check are sequential
+ * in the d_indices array:
+ * - tgt1_idx = d_indices[map_idx]   = 3 --> d_targets[3] == 'hi'
+ * - tgt2_idx = d_indices[map_idx+1] = 1 --> d_targets[1] == 'hello'
+ * The logic now only needs to check for either of these 2 targets.
+ *
+ * This kernel works in either row-per-string or warp-per-string depending
```

Comment:
Should this have been thread-per-string?
```suggestion
 * This kernel works in either thread-per-string or warp-per-string depending
```

Reply:
Yes, thread-per-string would more clear